mongorestore/restore.go

```diff
@@ -431,7 +431,7 @@ func (restore *MongoRestore) RestoreCollectionToDB(dbName, colName string,
 		totalResult.combineWith(<-resultChan)
 		if finalErr == nil && totalResult.Err != nil {
 			finalErr = totalResult.Err
-			close(restore.termChan)
+			restore.termChan <- struct{}{}
```

Comment:
Ah! I'm sorry, I suggested you do this but I just realized, this won't work. Multiple goroutines are listening to `restore.termChan`. If we send a single `struct{}{}`, only one of the goroutines will receive it. We do actually need to close the channel to ensure all the goroutines terminate.

So the correct thing to do is to check if the channel is already closed before closing it.

Reply:
> So the correct thing to do is to check if the channel is already closed before closing it.

Although this may end up being more complicated than necessary - there could be an easier way of doing this with some shared state and mutexes.

Reply:
Isn't checking the channel before closing it still racy? The channel could be closed by another goroutine after the check but before the call to close.

Another possibility (albeit a bit of a messy one) here would be to do the following:
- make `termChan` a buffered channel
- after a goroutine receives the signal on `termChan`, it writes that signal back onto the channel so that other goroutines can pick it up

Reply:
Yeah it's still racy. You'd have to use a mutex or a sync.Once to stop it being racy. 

I think using a channel makes it more complicated than it needs to be. We aren't using the channel for anything other than closing it. So we're essentially using it as a bool. I think the simplest thing to do would be to actually use a boolean value instead.

Reply:
Thanks for pointing this out! I was able to confirm that multiple go routines are listening to `restore.termChan` as the function `RestoreCollectionToDB` is called by `RestoreIntent` which is called by multiple go routines simultaneously. Thus, the most intuitive way to resolve this race condition is as Tim suggested -- change `restore.termChan` to a boolean ('restore.terminate'). Once `restore.terminate` is set, all go routines checking for the state of this variable will stop writing to their individual docChan's, finish reading the remaining data in docChan's buffer, and terminate with an error. `restore.terminate` is set in two cases -- (i) there is an error while attempting to restore a document or (ii) an interrupt signal is received. 

Since only one go routine attempts to close the local variable docChan per `RestoreCollectionToDB` thread, we will not encounter a double close panic here.